### PR TITLE
GFF3 rewrite

### LINKED ISSUE #371: Error showing value of type Bio.Intervals.GFF3Attributes
Not sure if GFF3 is parsing correctly.

I have GFF3 with the following line in it:

```
1 havana  processed_transcript  11869 14409 . + . ID=transcript:ENST00000456328;Parent=gene:ENSG00000223972;Name=DDX11L1-002;biotype=processed_transcript;havana_transcript=OTTHUMT00000362751;havana_version=1;tag=basic;transcript_id=ENST00000456328;transcript_support_level=1;version=2
```

I've opened up a GFF3 file. 

```julia
gff = GFF3Reader(open("/Shared/Au/bdeonovic/work/haplotype/gold_standard8/Homo_sapiens.GRCh38.85.n1000.gff3"));
line=read(gff)
```

I can take a look at the attributes of the current line:
```julia
for a in line.metadata.attributes
       println(a)
       end
"version"=>"transcript:ENST00000456328"
"tag"=>"2"
"biotype"=>"basic"
"havana_transcript"=>"processed_transcript"
"transcript_support_level"=>"OTTHUMT00000362751"
"havana_version"=>"1"
"transcript_id"=>"1"
"Name"=>"ENST00000456328"
"Parent"=>"DDX11L1-002"
"ID"=>"gene:ENSG00000223972"
"ID"=>"transcript:ENST00000456328"
```
the attribute names all seem a bit messed up to me. Also when I look at the attributes in REPL I get following error:

```julia
julia> line.metadata.attributes
Bio.Intervals.GFF3Attributes with 10 entries:
  Error showing value of type Bio.Intervals.GFF3Attributes:
ERROR: BoundsError: attempt to access 10-element Array{AbstractString,1} at index [11]
 in setindex!(::Array{AbstractString,1}, ::String, ::Int64) at ./array.jl:415
 in show(::IOContext{Base.Terminals.TTYTerminal}, ::MIME{Symbol("text/plain")}, ::Bio.Intervals.GFF3Attributes) at ./replutil.jl:64                                               
 in display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::MIME{Symbol("text/plain")}, ::Bio.Intervals.GFF3Attributes) at ./REPL.jl:132                                       
 in display(::Base.REPL.REPLDisplay{Base.REPL.LineEditREPL}, ::Bio.Intervals.GFF3Attributes) at ./REPL.jl:135                                                                     
 in display(::Bio.Intervals.GFF3Attributes) at ./multimedia.jl:143
 in print_response(::Base.Terminals.TTYTerminal, ::Any, ::Void, ::Bool, ::Bool, ::Void) at ./REPL.jl:154                                                                          
 in print_response(::Base.REPL.LineEditREPL, ::Any, ::Void, ::Bool, ::Bool) at ./REPL.jl:139                                                                                      
 in (::Base.REPL.##22#23{Bool,Base.REPL.##33#42{Base.REPL.LineEditREPL,Base.REPL.REPLHistoryProvider},Base.REPL.LineEditREPL,Base.LineEdit.Prompt})(::Base.LineEdit.MIState, ::Base.AbstractIOBuffer{Array{UInt8,1}}, ::Bool) at ./REPL.jl:652                             
 in run_interface(::Base.Terminals.TTYTerminal, ::Base.LineEdit.ModalInterface) at ./LineEdit.jl:1579                                                                             
 in run_frontend(::Base.REPL.LineEditREPL, ::Base.REPL.REPLBackendRef) at ./REPL.jl:903
 in run_repl(::Base.REPL.LineEditREPL, ::Base.##930#931) at ./REPL.jl:188
 in _start() at ./client.jl:360
```

### PR DESCRIPTION
**Pull request by bicycle1885**
This is a rewrite of the GFF3 parser including the ideas of #382 and #401.

Also, this will fix #371 (cc. @bdeonovic):
```
julia> record = GFF3.Record("1\thavana\tprocessed_transcript\t11869\t14409\t.\t+\t.\tID=transcript:ENST00000456328;Parent=gene:ENSG00000223972;Name=DDX11L1-002;biotype=processed_transcript;havana_transcript=OTTHUMT00000362751;havana_version=1;tag=basic;transcript_id=ENST00000456328;transcript_support_level=1;version=2")
Bio.Intervals.GFF3.Record:
    sequence: 1
      source: havana
        type: processed_transcript
       start: 11869
         end: 14409
       score: <missing>
      strand: +
       phase: <missing>
  attributes: ID=transcript:ENST00000456328 Parent=gene:ENSG00000223972 Name=DDX11L1-002 biotype=processed_transcript havana_transcript=OTTHUMT00000362751 havana_version=1 tag=basic transcript_id=ENST00000456328 transcript_support_level=1 version=2

julia> Dict(GFF3.attributes(record))
Dict{String,Array{String,1}} with 10 entries:
  "version"                  => String["2"]
  "tag"                      => String["basic"]
  "biotype"                  => String["processed_transcript"]
  "havana_transcript"        => String["OTTHUMT00000362751"]
  "transcript_support_level" => String["1"]
  "havana_version"           => String["1"]
  "transcript_id"            => String["ENST00000456328"]
  "Name"                     => String["DDX11L1-002"]
  "Parent"                   => String["gene:ENSG00000223972"]
  "ID"                       => String["transcript:ENST00000456328"]

```

- [x] Tests
- [x] Docs

### CODE REVIEW

**Comment by bicycle1885** [comment]
This is ready to review. The old code is kept for comparison but relocated under src/intervals/gff3_old. Comments are welcome.

File: src/ReaderHelper.jl [hunk 1]
```diff
@@ -52,7 +52,7 @@ function generate_index_function(record_type, machine, actions)
             cs = $(machine.start_state)
             $(Automa.generate_exec_code(machine, actions=actions, code=:goto, check=false))
             if cs != 0
-                throw(ArgumentError("failed to index $($(record_type))"))
+                throw(ArgumentError(string("failed to index $($(record_type))", " ~>", repr(String(data[p:min(p+7,p_end)])))))
```
Comment (Ward9250) [inline src/ReaderHelper.jl]: Why is it needed to interpolate `record_type` twice?

File: src/intervals/gff3/gff3.jl [hunk 2]
```diff
@@ -1,155 +1,17 @@
...
-
-function Base.copy(metadata::GFF3Metadata)
-    return GFF3Metadata(
-        copy(metadata.source),
-        copy(metadata.kind),
-        metadata.score,
-        metadata.phase,
-        copy(metadata.attributes))
-end
+include("record.jl")
```
Comment (Ward9250) [inline src/intervals/gff3/gff3.jl]: Yay for organisation into more specific files!!

File: src/intervals/gff3/reader.jl [hunk 3]
```diff
@@ -9,63 +11,57 @@ Create a reader for data in GFF3 format.
 * `save_directives=false`: if true, store directive lines, which can be accessed
   with the `directives` function
 """
-type GFF3Reader <: Bio.IO.AbstractReader
-    state::Ragel.State
-    version::VersionNumber
-    sequence_regions::Vector{Interval{Void}}
-    key::StringField
+type Reader <: Bio.IO.AbstractReader
+    state::Bio.Ragel.State
```
Comment (Ward9250) [inline src/intervals/gff3/reader.jl]: Didn't you change the Ragel stuff to be called something like Automa Helper? Does that need to be done here?

File: src/intervals/gff3/reader.jl [hunk 4]
```diff
@@ -9,63 +11,57 @@ Create a reader for data in GFF3 format.
...
+        if !skip_features
+            push!(targets, :feature)
+        end
+        if !skip_directives
+            push!(targets, :directive)
+        end
+        if !skip_comments
+            push!(targets, :comment)
+        end
+        return new(Bio.Ragel.State(body_machine.start_state, input), save_directives, targets, false, Record[], 0, 0)
```
Comment (Ward9250) [inline src/intervals/gff3/reader.jl]: Bio.Ragel.State - name change?

File: src/intervals/gff3/reader.jl [hunk 5]
```diff
@@ -95,9 +95,159 @@ Return a FASTAReader initialized to parse trailing FASTA data.
...
     return Bio.Seq.FASTAReader(reader.state.stream)
 end
+
+info("compiling GFF3")
+const record_machine, body_machine = (function ()
+    cat = Automa.RegExp.cat
+    rep = Automa.RegExp.rep
+    rep1 = Automa.RegExp.rep1
+    alt = Automa.RegExp.alt
+    opt = Automa.RegExp.opt
```
Comment (Ward9250) [inline src/intervals/gff3/reader.jl]: So I guess the purpose of this is if the machine definition is in an anonymous function, you know stuff like cat = Automa.RegExp.cat is contained only to the function, and is not causing unforseen consequences. If it's not already, we should include advice on this in the docs when telling people how to make their own parsers for new formats.

File: src/intervals/gff3/reader.jl (hunk 4 above)
Comment (bicycle1885) [inline src/intervals/gff3/reader.jl]: I will do, but not now.

File: src/intervals/gff3/reader.jl (hunk 5 above)
Comment (bicycle1885) [inline src/intervals/gff3/reader.jl]: Right. I will do that.

**Comment by bicycle1885** [comment]
I'm going to remove the old code and merge this today if no objections. I may add some follow-up commits here if needed.

**Comment by bicycle1885** [comment]
Oh, I need to address `IntervalCollection` stuffs before merging it.

**Comment by bicycle1885** [comment]
Travis failure on macOS is due to a bug of other package. I'm going to merge this in 12 hours if no further comments.

**Pull request merged**